src/core/decoder.js

```diff
@@ -58,75 +57,146 @@ function decodePolygon(geometry) {
+function decodeLine(geometry) {
     let vertices = [];
     let normals = [];
     let breakpoints = []; // Array of indices (to vertexArray) that separate each feature
-    geom.map(feature => {
+    geometry.map(feature => {
         feature.map(lineString => {
-            // Create triangulation
-
-            for (let i = 0; i < lineString.length - 2; i += 2) {
-                const a = [lineString[i + 0], lineString[i + 1]];
-                const b = [lineString[i + 2], lineString[i + 3]];
-                const normal = getLineNormal(b, a);
-                let na = normal;
-                let nb = normal;
+            addLine(lineString, vertices, normals);
+        });
+        breakpoints.push(vertices.length);    
+    });
+    return {
+        normals: new Float32Array(normals),
+        vertices: new Float32Array(vertices),
+        breakpoints
+    };
+}
 
-                if (i > 0) {
-                    const prev = [lineString[i - 2], lineString[i - 1]];
-                    na = getJointNormal(prev, a, b) || na;
-                }
-                if (i < lineString.length - 4) {
-                    const next = [lineString[i + 4], lineString[i + 5]];
-                    nb = getJointNormal(a, b, next) || nb;
+/**
+ * Create a triangulated lineString: zero-sized, vertex-shader expanded triangle list
+ * with `miter` joins. For angle < 60 joins are automatically adjusted to `bevel`.
+ */
+function addLine(lineString, vertices, normals) {
+    let prevPoint, currentPoint, nextPoint;
+    let prevNormal, nextNormal;
+    
+    // We need at least two points
+    if (lineString.length >= 4) {
+        
+        // Initialize the first two points
+        prevPoint = [lineString[0], lineString[1]];
+        currentPoint = [lineString[2], lineString[3]];
+        prevNormal = getLineNormal(prevPoint, currentPoint);
+        
+        for (let i = 4; i <= lineString.length; i += 2) {
+            
+            // First triangle
+            addTriangle(
+                [prevPoint, prevPoint, currentPoint],
+                [neg(prevNormal), prevNormal, prevNormal]
+            );
+            
+            // Second triangle
+            addTriangle(
+                [prevPoint, currentPoint, currentPoint],
+                [neg(prevNormal), prevNormal, neg(prevNormal)]
+            );
+            
+            if (i <= lineString.length - 2) {
+                // If there is a next point, compute its properties
+                nextPoint = [lineString[i], lineString[i + 1]];
+                nextNormal = getLineNormal(currentPoint, nextPoint);
+                // `turnLeft` indicates that the nextLine turns to the left
+                // `joinNormal` contains the direction and size for the `miter` vertex
+                //  If this is not defined means that the join must be `bevel`.
+                let {turnLeft, joinNormal} = getJoinNormal(prevNormal, nextNormal);
+                
+                // Third triangle
```

Comment:
I don't understand these two triangles, and how they are compared to the first two.
The change is quite important since previously we had 2 triangles with the joint normal. Now we are having two triangles with the line normal, another third triangle with a normal that I don't understand, and a conditional fourth triangle.

I guess these it has something to do with being the **interior** of the "curve" or being the exterior (by looking at `turnLeft`), but I don't get which triangle is each one.

Also, I'm worried about the performance of this, previously we were having 2 triangles per segment, now we have between 3 and 4.

Reply:
This is shown in the PR description. The third triangle is to define the `bevel` joins. The fourth triangle is used for the `miter` joins when they are required (that's why this is conditional).

The `turnLeft` is required to get the direction of the curve for having the correct orientation for the join's triangles.

In the previous implementation we had only 2 triangles per segment but with a bug. The current implementation gets rid of the bug and enables the feature for the `miter` and `bevel` style property. (Other implementation I have seen add up to 6 triangles per segment). And keep in mind that the `round` corner implementation adds lots of triangles for each join so this is not an issue.

As described in the PR, the first approach (with less number of triangles) has 2x worse performance because of the extra computations in JS and it doesn't solve the issue. The second approach has even less computations than the `master branch`'s algorithm and a similar performance in the shader side.